### DISCUSSION TITLE
template with multiple arguments

### DISCUSSION THREAD

**Discussion by daniel-torok** [hairyhenderson/gomplate#1176] [Q&A] [answered]
I found that we can have custom templates: https://docs.gomplate.ca/functions/tmpl/
The documentation also says that we can pass a `context` object to it, referenced as `{{ . }}` in the template.

Can we provide multiple arguments to a template?
For example:

```yaml
{{ define "Default" -}}
{{ if has .0 .1 }}{{ .0..1 }}{{ else }}{{ .2 }}{{ end }}
{{ end }}

{{ $ds := datasource "ds" }}
foo:
  bar: {{ template "Default" $ds keyToCheck defaultValue }}
```

In the example `.0` , `.1`, `.2` would refer to the indexed arguments the template received.

**Comment by hairyhenderson** [comment, marked as answer]
_(I've converted this to a discussion as this is more of a question than an issue)_

The context must be a single item, however that item can be a list (slice) or a map (dict). You can construct your own inline with the [`slice`](https://docs.gomplate.ca/functions/coll/#coll-slice) and [`dict`](https://docs.gomplate.ca/functions/coll/#coll-dict) functions:

```yaml
{{ define "foo" }}
{{ if has .ds .key }}{{ index .ds .key }}{{ else }}{{ .default }}{{ end }}
{{ end }}

{{ $ds := datasource "ds" }}
foo:
  bar: {{ template "foo" dict "ds" $ds "key" $keyToCheck "default" $defaultValue }}
```

(See [the docs](https://docs.gomplate.ca/syntax/#indexing-arrays-and-maps) for more about why `index` is necessary)

For more complex slices and maps it can be awkward to use the `slice` and `dict` functions extensively, so one pattern I've used a few times is to convert from inline YAML or JSON with the [`yaml`](https://docs.gomplate.ca/functions/data/#data-yaml) or [`json`](https://docs.gomplate.ca/functions/data/#data-json) functions. For example:

```yaml
{{ $ipRanges := `
internal:
- 10.0.0.0/8
- 172.16.0.0/12
- 192.168.0.0/24
external:
- 1.0.0.0/8
- 123.234.5.0/24
` | yaml }}

{{ template "someTemplate" $ipRanges }}
```

**Reply by daniel-torok** [reply]
Thank you, I didn't notice the `dict` and `slice` functions for this purpose.
